File: Java/src/main/java/TrainSorting.java
import java.util.Arrays;

public class TrainSorting {

    public static int solveProblem(int[] a) {
        if (a.length <= 2) return a.length;

        int[] lds = new int[a.length];
        int[] lis = new int[a.length];
        int res = 1;
        Arrays.fill(lds, 1);
        Arrays.fill(lis, 1);
        for (int i = a.length - 2; i >= 0; i--)
            for (int k = a.length - 1; k > i; k--) {
                if (a[k] < a[i] && lds[i] < lds[k] + 1)
                    lds[i] = lds[k] + 1;
                else if (a[k] > a[i] && lis[i] < lis[k] + 1)
                    lis[i] = lis[k] + 1;

                res = Math.max(res, (lis[i] + lds[i]) - 1);
            }
        return res;
    }

    public static void main(String[] args) {
        Kattio kattio = new Kattio(System.in, System.out);
        int n = kattio.getInt();

        int[] searchArray = new int[n];
        for (int i = 0; i < n; i++) searchArray[i] = kattio.getInt();

        System.out.println(solveProblem(searchArray));
    }
}

File: C++/src/TrainSorting.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

static int solveProblem(vector<int> *a, int n) {
    if (n <= 2) return n;

    vector<int> lds(n);
    vector<int> lis(n);
    int res = 1;

    fill(lds.begin(), lds.end(), 1);
    fill(lis.begin(), lis.end(), 1);

    for (int i = n - 2; i >= 0; i--)
        for (int k = n - 1; k > i; k--) {
            if (a->at(k) < a->at(i) && lds.at(i) < lds.at(k) + 1)
                lds.at(i) = lds.at(k) + 1;
            else if (a->at(k) > a->at(i) && lis.at(i) < lis.at(k) + 1)
                lis.at(i) = lis.at(k) + 1;

            res = max(res, ((lis.at(i) + lds.at(i)) - 1));
        }
    return res;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, s;
    cin >> n;

    auto *arr = new vector<int>();

    for (int i = 0; i < n; i++) {
        cin >> s;
        arr->push_back(s);
    }

    cout << solveProblem(arr, n) << endl;
}